src/Controller/HttpControllerAbstract.php

```diff
@@ -14,102 +17,72 @@
 /**
  * Base class for HTTP controllers.
  */
-abstract class HttpController extends ControllerAbstract
+abstract class HttpControllerAbstract extends ControllerAbstract
 {
     /**
-     * The request instance
-     *
-     * @var \YapepBase\Request\HttpRequest
+     * @var HttpRequest
      */
     protected $request;
 
     /**
-     * The response instance
-     *
-     * @var \YapepBase\Response\HttpResponse
+     * @var HttpResponse
      */
     protected $response;
 
-    /**
-     * Constructor.
-     *
-     * @param \YapepBase\Request\HttpRequest   $request  The request object. Must be a HttpRequest or descendant.
-     * @param \YapepBase\Response\HttpResponse $response The response object. Must be a HttpResponse or descendant.
-     *
-     * @throws \YapepBase\Exception\ControllerException   On error. (eg. incompatible request or response object)
-     */
-    public function __construct(IRequest $request, IResponse $response)
+    public function setRequest(IRequest $request)
     {
         if (!($request instanceof HttpRequest)) {
-            throw new ControllerException(
-                'The specified request is not a HttpRequest',
-                ControllerException::ERR_INCOMPATIBLE_REQUEST
-            );
+            throw new IncompatibleRequestException($request, HttpRequest::class);
         }
+
+        return parent::setRequest($request);
+    }
+
+    public function setResponse(IResponse $response)
+    {
         if (!($response instanceof HttpResponse)) {
-            throw new ControllerException(
-                'The specified response is not a HttpResponse',
-                ControllerException::ERR_INCOMPATIBLE_RESPONSE
-            );
+            throw new IncompatibleResponseException($response, HttpResponse::class);
         }
-        parent::__construct($request, $response);
+
+        return parent::setResponse($response);
     }
 
     /**
      * Redirects the client to the specified URL.
      *
-     * @param string $url        The URL to redirect to.
-     * @param int    $statusCode The status code of the redirect (3XX).
-     *
-     * @return void
-     *
-     * @throws \YapepBase\Exception\RedirectException   To stop execution of the controller.
+     * @throws RedirectException
      */
-    protected function redirectToUrl($url, $statusCode = 303)
+    protected function redirectToUrl(string $url, int $statusCode = 303): void
```

Comment:
Would be nice to use a constant for the status code. I think we already have a lib which lists all the status codes...

Reply:
I tried it in another project but it's a pain in the @ss.
The problem is that as a developer you always think of the code first when you use it, and you always need to find the name of the constant. 
Even bigger problem is that the IDE wont help you at all, as it orders the codecompletion list by the constant name not by the value obviously.
So even if you are familiar with the status codes and you know exactly what to use when, you will need to learn the constant names.